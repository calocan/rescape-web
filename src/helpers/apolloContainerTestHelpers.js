/**
 * Created by Andy Likuski on 2017.12.26
 * Copyright (c) 2017 Andy Likuski
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */


import {
  eitherToPromise, makeSampleInitialState, mockApolloClientWithSamples,
  propsFromSampleStateAndContainer, waitForChildComponentRender, wrapWithMockGraphqlAndStore
} from 'helpers/testHelpers';
import {getClass} from 'helpers/styleHelpers';
import * as R from 'ramda';
import * as Either from 'data.either';
import {theme} from 'styles/styles';
import {Provider as provider} from 'rebass';
import {eMap} from 'helpers/componentHelpers';
const [Provider] = eMap([provider])

/**
 * Runs tests on an apollo React container with the * given config.
 * Even if the container being tested does not have an apollo query, this can be used
 * @param {Object} config
 * @param {Object} config.container The React container to test
 * @param {String} config.componentName The name of the React component that the container wraps
 * @param {String} config.childClassDataName A class used in a React component in the named
 * component's renderData method--or any render code when apollo data is loaded
 * @param {Sting} config.childClassLoadingName A class used in a React component in the named
 * component's renderLoading method--or any render code called when apollo loading is true
 * @param {Sting} config.childClassErrorName A class used in a React component in the named
 * component's renderError method--or any render code called when apollo error is true
 * @param {Function} config.testPropsMaker A function defined in the conatiner being tested that is
 * either created with makeApolloTestPropsFunction or makeTestPropsFunction, the former
 * if the container has an Apollo query and the latter if it doesn't. This function
 * and the result of asyncParentProps is passed propsFromSampleStateAndContainer, which
 * ultimately generates the test props
 * @param {Function} config.asyncParentProps A function with no arguments that returns a Promise
 * of the parentProps used to call propsFromSampleStateAndContainer
 * @param {Object} [config.query] Optional gql wrapped query string if the Container has an apollo query.
 * The query should be the same as that used by the container
 * @param {Function} [config.queryVariables] Optional Unary function expecting props and return an object of query arguments
 * for the given query. Example:
 * props => ({
    regionId: props.data.region.id
  });
 * @param {Function} [config.errorMaker] Optional unary function that expects the results of the
 * parentProps and mutates something used by the queryVariables to make the query fail. This
 * is for testing the renderError part of the component
 */
export const apolloContainerTests = (config) => {

    const {
      Container,
      componentName,
      childClassDataName,
      // Optional, the class name if the component has an Apollo-based loading state
      childClassLoadingName,
      // Optional, the class name if the component has an Apollo-based error state
      childClassErrorName,
      // Optional, if not specified then no container props are needed
      testPropsMaker,
      // Optional, must return parent props as a Promise
      asyncParentProps,

      // Optional. Only for components with queries
      query,
      // Optional. Only for components with queries
      queryVariables,
      // Optional. Only for components with queries
      errorMaker
    } = config;

    // Get the test props for this container
    const asyncProps = () =>
      asyncParentProps ?
        asyncParentProps()
          .then(parentProps => {
            const result = testPropsMaker ? propsFromSampleStateAndContainer(testPropsMaker, parentProps) : parentProps;
            return R.unless(
              R.is(Promise),
              res => Promise.resolve(Either.Right(res))
            )(result);
          })
          .then(eitherToPromise) :
        Promise.resolve({});

    const asyncParentPropsOrDefault = asyncParentProps ? asyncParentProps() : Promise.resolve({});

    test('mapStateToProps', async () => {
      // Get the test props for RegionContainer
      const props = await asyncProps();
      expect(props).toMatchSnapshot();
    });

    if (query) {
      test('query', async () => {
        const parentProps = await asyncParentPropsOrDefault;
        const props = await propsFromSampleStateAndContainer(testPropsMaker, parentProps).then(eitherToPromise);
        const data = await mockApolloClientWithSamples().query({
          query,
          variables: queryVariables(props),
          context: {
            dataSource: makeSampleInitialState()
          }
        });
        expect(data).toMatchSnapshot();
      });
    }

    test('render', async (done) => {
      // Wait for the sample parent props that feed this component
      const parentProps = await asyncParentPropsOrDefault;
      // Wrap the component in mock Apollo and Redux providers.
      // If the compoent doesn't use Apollo it just means that it will render its children synchronously,
      // rather than asynchronously
      const wrapper = wrapWithMockGraphqlAndStore(Provider({theme}, Container(parentProps)));
      // Find the top-level component. This is always rendered in any Apollo status (loading, error, store data)
      const component = wrapper.find(componentName);
      // Make sure the component props are consistent since the last test run
      expect(component.props()).toMatchSnapshot();

      // If we have an Apollo component, our immediate status after mounting the component is loading. Confirm
      if (childClassLoadingName) {
        expect(component.find(`.${getClass(childClassLoadingName)}`).length).toEqual(1);
      }

      // If we have an Apollo component, we use enzyme-wait to await the query to run and the the child
      // component that is dependent on the query result to render. If we don't have an Apollo component,
      // this child will be rendered immediately without delay
      waitForChildComponentRender(wrapper, componentName, childClassDataName).then(
        childComponent => {
          expect(childComponent.props()).toMatchSnapshot();
          done();
        }
      ).catch(e => {
        // If we don't find it a timeout error or other error occured.
        throw e;
        done();
      });
    });

    if (errorMaker) {
      test('renderError', async (done) => {
        const parentProps = await asyncParentPropsOrDefault.then(errorMaker);
        const wrapper = wrapWithMockGraphqlAndStore(Container(parentProps));
        const component = wrapper.find(componentName);
        expect(component.find(`.${getClass(childClassLoadingName)}`).length).toEqual(1);
        expect(component.props()).toMatchSnapshot();
        waitForChildComponentRender(wrapper, componentName, childClassErrorName, done);
      });
    }
  }
;
