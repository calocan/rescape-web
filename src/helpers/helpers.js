/**
 * Created by Andy Likuski on 2018.01.17
 * Copyright (c) 2018 Andy Likuski
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {createSampleConfig, getCurrentConfig} from 'rescape-sample-data';
import makeSchema from 'schema/schema';
import {
  makeApolloTestPropsTaskFunction, makeSampleInitialState,
  propsFromSampleStateAndContainer,
  makeTestPropsFunction
} from 'rescape-helpers-component';
import {createSelectorResolvedSchema} from 'schema/selectorResolvers';
import createInitialState from 'initialState';
import {promiseToTask, taskToPromise, reqPathThrowing} from 'rescape-ramda';
import {of} from 'folktale/concurrency/task';
import PropTypes from 'prop-types';
import {v} from 'rescape-validate';
import * as Either from 'data.either';

/**
 *
 * These helpers are trivial functions that use rescape-sample-data and rescape-helpers-component togther,
 * creating functions that are seeded with sample data for testing
 */

const sampleConfig = createSampleConfig();
/**
 * Sample initial state for tests
 */
export const sampleInitialState = makeSampleInitialState(createInitialState, sampleConfig);

/**
 * Resolved schema for tests
 * @type {Object}
 */
export const resolvedSchema = createSelectorResolvedSchema(makeSchema(), getCurrentConfig());

/**
 * Calls makenApolloTestPropsFunction with the given * 3 arguments:
 * mapStateToProps, mapDispatchToProps, {query, args}
 * See makeApolloTestPropsFunction for details.
 * The resulting promise is wrapped in a Task.
 * @param {Function} mapStateToProps
 * @param {Function} mapDispatchToProps
 * @param {Object} queryInfo
 * @returns {Function} A 2 arity function called with state and props that results in a Task that
 * resolves the props as A Right if no errors and a Left if errors
 */
export const apolloTestPropsTaskMaker = v((mapStateToProps, mapDispatchToProps, queryInfo) =>
    (state, props) => makeApolloTestPropsTaskFunction(resolvedSchema, sampleConfig, mapStateToProps, mapDispatchToProps, queryInfo)(state, props)
  , [
    ['mapStateToProps', PropTypes.func.isRequired],
    ['mapDispatchToProps', PropTypes.func.isRequired],
    ['queryInfo', PropTypes.shape().isRequired]
  ], 'apolloTestPropsTaskMaker');

/**
 * Calls makeTestPropsFunction on a non Apollo container. This is a synchronous but wrapped in a
 * Task to match calls to apolloTestPropsTaskMaker
 * @param mapStateToProps
 * @param mapDispatchToProps
 * @return {Function} A 2 arity function called with state and props that results in a Task that
 * resolves the props
 */
export const testPropsTaskMaker = (mapStateToProps, mapDispatchToProps) =>
  // Wrap function result in a Task to match apolloTestPropsTaskMaker
  (state, props) => of(Either.Right(makeTestPropsFunction(mapStateToProps, mapDispatchToProps)(state, props)));

/**
 * Given a Task to fetch parent container props and a task to fetch the current container props,
 * Fetches the parent props and then samplePropsTaskMaker with the initial state and parent props
 * @param {Task} parentContainerSamplePropsTask Task that resolves to the parent container props
 * @param {Function} samplePropsTaskMaker 2 arity function expecting state and parent props.
 * Returns a Task from a container that expects a sample state and sampleOwnProps
 * and then applies the container's mapStateToProps, mapDispatchToProps, and optional mergeProps
 * @param parentComponentViews
 * @param viewName
 * @returns {Task} A Task to asynchronously return the parentContainer props merged with sampleOwnProps
 * in an Either.Right. If anything goes wrong an Either.Left is returned
 */
export const asyncParentPropsTask = v((parentContainerSamplePropsTask, samplePropsTaskMaker, parentComponentViews, viewName) =>
    parentContainerSamplePropsTask.map(
      // the parent props to the props of the desired view in an Either.Right
      propsEither => propsEither.map(props => reqPathThrowing(['views', viewName], parentComponentViews(props)))
    ).chain(parentContainerSamplePropsEither => parentContainerSamplePropsEither
      // Chain the Either.Right value to a Task combine the parent props with the props maker
        .chain(parentContainerSampleProps => samplePropsTaskMaker(sampleInitialState, parentContainerSampleProps))
    ).map(value => value),
  [
    ['parentContainerSamplePropsTask', PropTypes.shape().isRequired],
    ['samplePropsTaskMaker', PropTypes.func.isRequired],
    ['parentComponentViews', PropTypes.func.isRequired],
    ['viewName', PropTypes.string.isRequired]
  ],
  'asyncParentPropsTask');


