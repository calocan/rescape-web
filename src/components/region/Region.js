/**
 * Created by Andy Likuski on 2017.02.16
 * Copyright (c) 2017 Andy Likuski
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
import {styleMultiplier} from 'helpers/styleHelpers';
import mapbox from 'components/map/mapbox/MapboxContainer';
import sankey from 'components/map/sankey/SankeyContainer';
import markerList from 'components/map/marker/MarkerListContainer';
import PropTypes from 'prop-types';
import {applyMatchingStyles, mergeAndApplyMatchingStyles} from 'selectors/styleSelectors';
import {
  nameLookup, eMap, propsFor, renderChoicepoint, composeViews,
  propsForSansClass, renderLoadingDefault, renderErrorDefault
} from 'helpers/componentHelpers';
import {mergeDeep, strPath, throwing} from 'rescape-ramda';
import * as R from 'ramda';
import {Component} from 'react'
const {reqPath} = throwing

const [Mapbox, Sankey, MarkerList, Div] = eMap([mapbox, sankey, markerList, 'div']);
export const c = nameLookup({
  region: true,
  regionMapboxOuter: true,
  regionMapbox: true,
  regionSankey: true,
  regionLocationsOuter: true,
  regionLocations: true,
  regionLoading: true,
  regionError: true
});

/**
 * The View for a Region, such as California. Theoretically we could display multiple regions at once
 * if we had more than one, or we could have a zoomed in region of California like the Bay Area.
 */
export default class Region extends Component {
  render() {
    const props = Region.views(this.props);
    return Div(propsFor(props.views, c.region),
      Region.choicepoint(props)
    );
  }
}

/**
 * Merges parent and state styles into component styles
 * @param style
 */
Region.viewStyles = ({style}) => {
  return {
    [c.region]: mergeAndApplyMatchingStyles(style, {
      position: 'absolute',
      width: styleMultiplier(1),
      height: styleMultiplier(1)
    }),

    [c.regionMapboxOuter]: applyMatchingStyles(style, {
      position: 'absolute',
      width: styleMultiplier(.5),
      height: styleMultiplier(1)
    }),

    [c.regionMapbox]: R.merge(
      // Pass width and height to Mapbox component
      R.pick(['width', 'height'], style),
      {
        // Other styles to pass to component (unlikely)
      }
    ),

    [c.regionSankey]: R.merge(
      // Pass width and height to Mapbox component
      R.pick(['width', 'height'], style),
      {
        // Other styles to pass to component (unlikely)
      }
    ),

    [c.regionLocationsOuter]: {
      position: 'absolute',
      top: .02,
      left: .55,
      right: .05
    }
  }
};

Region.viewProps = () => {
  // region is expected from the query result
  const region = strPath('data.store.region');
  return {
    [c.region]: {region},
    [c.regionMapbox]: {region},
    [c.regionMapboxOuter]: {key: c.regionMapboxOuter},
    [c.regionSankey]: {region},
    [c.regionLocationsOuter]: {key: c.regionLocationsOuter}
  };
};

Region.viewActions = () => {
  return {
    [c.regionMapbox]: {}
  };
};

Region.renderData = ({views}) => {
  const props = propsFor(views);
  const propsSansClass = propsForSansClass(views);

  return [
    Div(props(c.regionMapboxOuter),
      //Mapbox(
      //  propsSansClass(c.regionMapbox)
      //),
      Sankey(
        propsSansClass(c.regionSankey)
      )
    ),
    //Div(props(c.regionLocationsOuter),
    //  MarkerList(propsSansClass(c.regionLocations))
    //)
  ];
};


/**
 * Adds to props.views for each component configured in viewActions, viewProps, and viewStyles
 * @param {Object} props this.props or equivalent for testing
 * @returns {Object} modified props
 */
Region.views = composeViews(
  Region.viewActions(),
  Region.viewProps(),
  Region.viewStyles
);

/**
 * Loading, Error, or Data based on the props
 */
Region.choicepoint = renderChoicepoint(
  renderErrorDefault(c.regionError),
  renderLoadingDefault(c.regionLoading),
  Region.renderData
);

Region.propTypes = {
  data: PropTypes.shape().isRequired,
  style: PropTypes.shape().isRequired
};

Region.propTypes = {};
