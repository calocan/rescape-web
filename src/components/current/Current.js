/**
 * Created by Andy Likuski on 2016.05.26
 * Copyright (c) 2016 Andy Likuski
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import PropTypes from 'prop-types';
import region from 'components/region/RegionContainer';
import styles from './Current.style';
import React from 'react';
import {eMap} from 'helpers/componentHelpers';
import * as R from 'ramda';
import prettyFormat from 'pretty-format';

const [Div, Region] =
  eMap(['div', region]);

/**
 * Displays the Region of the current state and eventually a Region selector.
 * This might also be modified to display all available regions, perhaps on a continental map
 */
class Current extends React.Component {
  render() {
    return R.cond([
      [R.propEq('loading'), () => this.renderLoading()],
      [R.has('error'), error => this.renderError(error)],
      [R.has('data'), data => this.renderData(data)],
      [R.T, () => this.renderError(new Error("Expected loading, error, or data from Apollo wrapper, didn't get anything"))]
    ])(this.props);
  }

  renderLoading() {
    return Div();
  }

  renderError(error) {
    return Div();
  }

  renderData({region}) {
    return Div(
      {className: 'current'},
      Region({
          region,
          style: {
            width: this.props.style.width,
            height: this.props.style.height
          }
        },
        null
      )
    );
  }
}

const {
  number,
  shape
} = PropTypes;

/**
 * Expect the current region
 * @type {{region: *}}
 */
Current.propTypes = {
  style: shape({
    width: number.isRequired,
    height: number.isRequired
  })
};

export default Current;
