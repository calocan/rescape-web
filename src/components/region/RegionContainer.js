
import {connect} from 'react-redux';
import Region from './Region'
import {makeMergeDefaultStyleWithProps} from 'selectors/styleSelectors';
import {createSelector} from 'reselect';
import * as R from 'ramda';
import {gql} from 'apollo-client-preset';
import {graphql} from 'react-apollo';

/**
 * RegionContainer expects the state to contain the active user and that user's Regions
 * It is given one of those regions as props
 * It calculates the values needed by Region's child components
 * @param {Object} state The redux state
 * @param {Object} props The parent props
 * @param {Object} props.Region The region
 */
export const mapStateToProps =
  (state, props) => createSelector(
    [
      makeMergeDefaultStyleWithProps(),
    ],
    style => {
      return R.mergeAll([
        props,
        {style},
      ]);
    }
  )(state, props);

export const mapDispatchToProps = (dispatch) => {
  return {
    /*
    onRegionIsChanged: (options, bounds) => {
      dispatch({
        type: actions.FETCH_TRANSIT_DATA,
        args: [options, bounds]
      });
    }
    */
  };
};


/**
 * Query
 * Prerequisites:
 *   A User in context
 * Resolves:
 *  The Regions of the User
 * Without prerequisites:
 *  Skip render
 */
const query = gql`
    query region($regionId: String!) {
        store {
            region(id: $id) {
                id
                name
            },
        }
    }
`;

const ContainerWithData = graphql(query, {
  options: ({region}) => ({
    variables: {
      regionId: region.id
    }
  }),
  props: ({data}) => ({
    store: data.store,
    loading: data.loading,
    data
  })
})
(Region);

const RegionContainer = connect(
  /**
   * The wrapped component needs access to the settings and a r
   * @param state
   * @returns {{}}
   */
  mapStateToProps, mapDispatchToProps
)(ContainerWithData);

export default RegionContainer;

