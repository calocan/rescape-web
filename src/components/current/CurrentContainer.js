import {connect} from 'react-redux';
import Current from './Current'
import * as R from 'ramda';
import {createSelector} from 'reselect';
import { makeActiveUserAndSettingsStateSelector } from 'selectors/storeSelectors';
import {makeBrowserProportionalDimensionsSelector} from 'selectors/styleSelectors';
import {mergeDeep, throwing} from 'rescape-ramda'
import {makeTestPropsFunction, mergePropsForViews} from 'helpers/componentHelpers';
import {bindActionCreators} from 'redux';
import {activeUserRegionSelector} from 'selectors/userSelectors';

/**
 * Combined selector that:
 * Limits the state to the active user and region. This merges props with the state immediately, overriding the
 * state with anything in props.
 * It does not pass props on to sub selectors as a separate argument
 * Limits the component dimensions to the browser.
 * @param {Object} state The redux state
 * @param {Object} [props] The optional props to override the state.
 * @returns {Object} The state and own props mapped to props for the component
 */
export const mapStateToProps = (state, props) =>
  createSelector(
    [
      (state, props) => {
        return makeActiveUserAndSettingsStateSelector()(mergeDeep(state, R.defaultTo({}, props)));
      },
      makeBrowserProportionalDimensionsSelector()
    ],
    (activeUserAndSettings, style) =>
      ({
        // No current graphql queries, pass the winnowed state
        // It might turn out that Current doesn't need anything because it simply renders child containers
        // Merge the browser dimensions with the props
        // props from the parent contain style instructions
        // TODO we need to set width and height proportional to the browser dimensions, not equal to
        data: mergeDeep(activeUserAndSettings, {style}),
        views: {
          regionProps: {
            region: activeUserRegionSelector(activeUserAndSettings),
            style: {
              width: style.width,
              height: style.height
            }
          }
        }
      })
  )(state, props);

export const mapDispatchToProps = (dispatch, ownProps) => {
  return bindActionCreators({
  }, dispatch)
}

/**
 * Combines mapStateToProps, mapDispatchToProps with the given viewToActions mapping
 * @type {Function}
 */
export const mergeProps = mergePropsForViews({
  // Region child component needs the following actions
  region: []
})

// Returns a function that expects ownProps for testing
export const testPropsMaker = makeTestPropsFunction(mapStateToProps, mapDispatchToProps, mergeProps)

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(Current)
