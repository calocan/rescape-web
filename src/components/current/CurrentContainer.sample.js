/**
 * Created by Andy Likuski on 2018.03.07
 * Copyright (c) 2018 Andy Likuski
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {makeTestPropsFunction, propsFromSampleStateAndContainer} from 'rescape-helpers-component';
import {mapDispatchToProps, mapStateToProps} from 'components/current/CurrentContainer';
import {sampleInitialState} from 'helpers/helpers';
import Current, {c} from './Current';
import {reqStrPathThrowing} from 'rescape-ramda';

/**
 * Returns a function that expects ownProps for testing
 */
export const samplePropsMaker = makeTestPropsFunction(mapStateToProps, mapDispatchToProps);

/**
 * Returns a function that when called returns a promise of sample parentProps
 * @param {String} viewName The view of the parent component representing the child Component whose props are sought
 * @return {function(): Promise<any>}
 */
export const sampleAsyncParentProps = (viewName) => () => new Promise((resolve) => {
  const currentProps = propsFromSampleStateAndContainer(sampleInitialState, samplePropsMaker, {});
  const parentProps = reqStrPathThrowing(viewName, Current.views(currentProps).views);
  resolve(parentProps);
});