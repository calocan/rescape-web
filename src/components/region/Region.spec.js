import React from 'react';

import {propsFromSampleStateAndContainer, shallowWrap} from 'helpers/testHelpers';
import region from './Region'
import {eMap} from 'helpers/componentHelpers';
import {testPropsMaker} from 'components/current/CurrentContainer';
const [Region] = eMap([region]);

describe('Region', () => {
  const props = propsFromSampleStateAndContainer(testPropsMaker,
    {
      // style dimensions are normally from the parent
      style: {
        width: 500,
        height: 500
      }
    }
  );

  test('Current can mount', () => {
    expect(shallowWrap(Region, props)).toMatchSnapshot();
  });
});
