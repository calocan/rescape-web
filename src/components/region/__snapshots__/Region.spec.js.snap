// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Region Can mount 1`] = `
<div
  className="region-root"
  style={
    Object {
      "height": 500,
      "position": "absolute",
      "width": 500,
    }
  }
>
  <div
    className="region-mapbox-outer"
    style={Object {}}
  >
    <Connect(Sankey)
      style={
        Object {
          "height": 500,
          "position": "absolute",
          "width": 250,
        }
      }
    />
  </div>
  <div
    className="region-markers-outer"
    style={
      Object {
        "left": 0.55,
        "position": "absolute",
        "right": 0.05,
        "top": 0.02,
      }
    }
  >
    <Connect(MarkerList) />
  </div>
</div>
`;

exports[`Region Current can mount 1`] = `
<div
  className={[Function]}
  style={
    Object {
      "height": [Function],
      "position": "absolute",
      "width": [Function],
    }
  }
>
  <div
    className={[Function]}
    style={Object {}}
  >
    <Connect(Mapbox)
      style={
        Object {
          "height": [Function],
          "position": "absolute",
          "width": [Function],
        }
      }
    />
    <Connect(Sankey)
      style={
        Object {
          "height": [Function],
          "position": "absolute",
          "width": [Function],
        }
      }
    />
  </div>
  <div
    className={[Function]}
    style={
      Object {
        "left": 0.55,
        "position": "absolute",
        "right": 0.05,
        "top": 0.02,
      }
    }
  >
    <Connect(MarkerList) />
  </div>
</div>
`;
